### ISSUE TITLE
Possible bug for bismark when methylation calling ?

### ISSUE THREAD

**Issue by hxlei** [FelixKrueger/Bismark#69]
I'm using bismark for bs-seq mapping and methylation calling. I take a glance at the coverage file generated by bismark2bedGraph. However, the result seems strange. Take CpG site of chrM for example (because chrM is the smallest chromosome). The coverage file contains sites that are not CpG sites. Here is a part of one CpG coverage file.
---

chrM 96 96 0 0 95
chrM 97 97 0 0 92
chrM 100 100 0 0 1
chrM 105 105 0 0 92
chrM 106 106 0 0 93
... ...
chrM 2003 2003 0 0 170
chrM 2004 2004 0 0 116
chrM 2107 2107 0 0 1
chrM 2204 2204 0 0 163
chrM 2205 2205 0 0 156
... ...
chrM 2826 2826 0 0 125
chrM 2827 2827 0 0 147
chrM 2833 2833 0 0 1
chrM 2845 2845 0 0 130
chrM 2846 2846 0 0 142
---

It's strange that site 100,2107,2833 are mixed in the CpG coverage file. And on the same sites of the reference, these three sites are not CpG sites. (here I call it false CpG.)
Then I count total CpG sites on chrM (hg19). There are 439 CpG sites on the plus strand, whick means 878 CpG on both strands. 
So number of lines in chrM CpG coverage file should always be equal to or less than 878. But the truth is, for two bs-seq data, one is 885, another is 900. Both of them contains false CpG. And the sites are different between them. For example, site 100 in the first file doesn't appear in the second one. Another, all of the false CpG only have one unmethylated reads.

Here are my commands (%s represents parameters):
`bismark --multicore %s --output_dir %s %s -1 %s -2 %s
bismark_methylation_extractor -p --no_overlap --comprehensive --multicore %s --ignore %s --ignore_r2 %s --ignore_3prime %s --ignore_3prime_r2 %s --output  %s %s
bismark2bedGraph --dir %s --remove_spaces --CX --ample_memory --output CpG %s`

First, I use --comprehensive to produce CpG,CHH and CHG context output files. Then I use bismark2bedGraph and --CX to process CpG,CHH and CHG context output files separately. In this way I can get CpG,CHH and CHG coverage files separately.
In addition, my version is v0.16.1. I haven't test it using the newest one.

I dont know wether I can use the final results to continue analysis.

Many thanks.

**Comment by FelixKrueger** [comment]
Hi hxlei,

This is a question I have seen a number of times, and I believe I have got an answer for your. The coverage files contain every position that received a CpG methylation call during the mapping step. There are certain cases where you can have changes of cytosine context due to deletions in the immediate proximity, like in such a scenario:
 
`CAATGGGA`  where the first C would be in `CHH` context. If there was a deletion of `AAT`, you would get
 
`C---GGGA`, and the context would effectively have changed from `CHH` to `CG`. It is quite likely such a change would indeed also affect the methylation state which is why these positions are not discarded altogether. This really only happens for deletions in the immediate proximity of a cytosine but not insertions, simply because insertions are padded with `N` during the methylation call process which would render them `Unknown` context.
 
The script `coverage2cytosine` on the other hand is reference genome based, so it will go through the reference sequence and check whether a CpG was covered or not. It will therefore ‘miss’ out any of the calls where a deletion had changed the sequence context. 

So bottom line: we believe it should be fine to proceed with the default behavior of calling changed sequence context because it is very likely that a change from CHH to CpG or vice versa will really also lead to a change in that residue's methylation state. If you wanted to exclude then anyway then you might continue with `coverage2cytosine`. I hope this makes sense?

NB: I would really recommend upgrading to the latest version 0.16.3. because of a known bug with ambiguous alignments in 0.16.1 (see here: a831ce73c22ef5b60fdc8f3f58b462417ddb59a2 and 966f3e3e4e09167dd82ac1dc5bd30eedd04f7f8f).

Best wishes, Felix
